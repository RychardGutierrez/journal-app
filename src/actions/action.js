export const newAction = (type, payload) => ({ type, payload });
