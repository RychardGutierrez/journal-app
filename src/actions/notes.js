import { deleteDoc, doc, updateDoc } from 'firebase/firestore';
import {
  fireStoreAddDoc,
  fireStoreCollection,
  fireStoreDataBase,
} from '../firebase/firebase-config';

import { NOTE_PATH, NEW_NOTE } from '../common/constants';
import { loadNotes } from '../common/loadNotes';
import { imgFileUpload } from '../common/utils';
import { types } from '../types';
import { newAction } from './action';
import {
  messageClose,
  messageNoteDeleteWait,
  messageNoteUpdate,
} from '../common/messages';
import { messageNoteUpdateError } from './../common/errorMessage';
import { messageNoteUpdateWait } from './../common/messages';

export const startNewNote = () => async (dispatch, getState) => {
  try {
    const { uid } = getState().auth;
    const documentRef = await fireStoreAddDoc(
      fireStoreCollection(fireStoreDataBase, `${uid}${NOTE_PATH}`),
      NEW_NOTE
    );

    dispatch(activeNote(documentRef.id, NEW_NOTE));
    dispatch(startLoadingNotes(uid));
  } catch (error) {
    console.log(error);
  }
};

export const activeNote = (id, note) =>
  newAction(types.notesActive, { ...note, id });

export const startLoadingNotes = (uid) => async (dispatch) => {
  const notes = await loadNotes(uid);
  dispatch(newAction(types.notesLoad, notes));
};

export const startLoadingEntry = (id) => (dispatch, getState) => {
  const { data } = getState().notes;
  dispatch(activeNote(id, data[id]));
};

export const refreshNote = (id, note) =>
  newAction(types.notesUpdated, { ...note, id });

export const startSaveNote = (note) => async (dispatch, getState) => {
  try {
    const { uid } = getState().auth;
    const { active } = getState().notes;
    const noteReference = doc(
      fireStoreDataBase,
      `${uid}${NOTE_PATH}/${active.id}`
    );
    const noteToFireStore = { ...note };

    !noteToFireStore.url && delete noteToFireStore.url;
    delete noteToFireStore.id;
    await updateDoc(noteReference, noteToFireStore);
    dispatch(refreshNote(note.id, noteToFireStore));

    messageNoteUpdate();
  } catch (error) {
    messageNoteUpdateError();
  }
};

export const startUploadingImg = (imgFile) => async (dispatch, getState) => {
  messageNoteUpdateWait();

  const { active } = getState().notes;
  const fileUrl = await imgFileUpload(imgFile);
  const newNoteActive = { ...active, url: fileUrl };
  dispatch(startSaveNote(newNoteActive));

  messageClose();
};

const deleteNote = (id) => newAction(types.notesDelete, { id });

export const startDeleteNote = (id) => async (dispatch, getState) => {
  try {
    messageNoteDeleteWait();

    const { uid } = getState().auth;
    const noteRef = doc(fireStoreDataBase, `${uid}${NOTE_PATH}/${id}`);
    await deleteDoc(noteRef);
    dispatch(deleteNote(id));

    messageClose();
  } catch (error) {
    console.log(error);
  }
};

export const cleanNote = () => newAction(types.notesLogoutClean);
