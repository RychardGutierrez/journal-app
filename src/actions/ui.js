import { types } from "../types";
import { newAction } from "./action";

export const uiSetErrorMessage = ({ messageError }) =>
  newAction(types.uiSetError, { messageError });

export const uiRemoveErrorMessage = () => newAction(types.uiRemoveError);

export const uiStartLoading = () => newAction(types.uiStartLoading);

export const uiEndLoading = () => newAction(types.uiEndLoading);
