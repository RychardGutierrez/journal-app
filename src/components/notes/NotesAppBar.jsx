import { useRef } from "react";
import PropTypes from "prop-types";
import { useDispatch } from "react-redux";
import dayjs from "dayjs";

import { startUploadingImg } from "../../actions/notes";
import { FORMAT_DATE, PICTURE, SAVE } from "../../common/constants";

const NotesAppBar = ({ createdDate, handleSaveNote }) => {
  const dispatch = useDispatch();
  const refInputImg = useRef(null);

  const dateNote = dayjs(createdDate).format(FORMAT_DATE);

  const handleUploadPicture = () => {
    refInputImg.current.click();
  };

  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (!!file) {
      dispatch(startUploadingImg(file));
    }
  };

  return (
    <div className="notes__appbar">
      <span>{dateNote}</span>
      <input
        type="file"
        className="notes__file"
        onChange={handleFileChange}
        ref={refInputImg}
      />
      <div>
        <button className="btn" onClick={handleUploadPicture}>
          {PICTURE}
        </button>
        <button className="btn" onClick={handleSaveNote}>
          {SAVE}
        </button>
      </div>
    </div>
  );
};

NotesAppBar.prototype = {
  createdDate: PropTypes.string.isRequired,
  handleSaveNote: PropTypes.func.isRequired,
};

export default NotesAppBar;
