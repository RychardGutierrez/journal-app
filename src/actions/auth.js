import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword as createUserFirebase,
  updateProfile,
  signInWithEmailAndPassword as signInUserFirebase,
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';

import { types } from '../types';
import { newAction } from './action';
import { cleanNote } from './notes';
import { uiEndLoading, uiSetErrorMessage, uiStartLoading } from './ui';
import {
  messageErrorRegisterNewUser,
  messageLoginError,
} from '../common/errorMessage';
import { messageWelcome } from '../common/messages';

export const login = (uid, name) => newAction(types.login, { uid, name });

export const startLogout = () => async (dispatch) => {
  try {
    await getAuth().signOut();
    dispatch(newAction(types.logout));
    dispatch(cleanNote());
  } catch (error) {
    console.log(error);
  }
};

export const startLoginEmailPassword =
  ({ email, password }) =>
  async (dispatch) => {
    dispatch(uiStartLoading());
    try {
      const { user } = await signInUserFirebase(getAuth(), email, password);
      const response = dispatch(login(user.uid, user.displayName));

      messageWelcome(user.displayName);
      return response;
    } catch (error) {
      return messageLoginError();
    } finally {
      return dispatch(uiEndLoading());
    }
  };

export const startGoogleLogin = () => async (dispatch) => {
  try {
    const { user } = await signInWithPopup(getAuth(), googleAuthProvider);
    dispatch(login(user.uid, user.displayName));

    messageWelcome(user.displayName);
  } catch (error) {
    messageLoginError();
  } finally {
    dispatch(uiEndLoading());
  }
};

export const startRegisterWithEmailPasswordName =
  ({ email, password, name }) =>
  async (dispatch) => {
    dispatch(uiStartLoading());
    try {
      const { user } = await createUserFirebase(getAuth(), email, password);
      await updateProfile(user, { displayName: name });
      dispatch(login(user.uid, user.displayName));

      messageWelcome(user.displayName);
    } catch (error) {
      dispatch(uiSetErrorMessage(messageErrorRegisterNewUser));
    } finally {
      dispatch(uiEndLoading());
    }
  };
