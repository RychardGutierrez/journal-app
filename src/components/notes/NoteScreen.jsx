import { useEffect, useMemo, useRef } from "react";
import PropTypes from "prop-types";
import { useDispatch } from "react-redux";
import { activeNote, startSaveNote } from "../../actions/notes";

import { useForm } from "../../hooks/useForm";
import NotesAppBar from "./NotesAppBar";
import { startDeleteNote } from "./../../actions/notes";
import { messageDelete } from "../../common/messages";
import { DELETE } from "../../common/constants";

const NoteScreen = ({ id, title, body, url, createdDate }) => {
  const dispatch = useDispatch();
  const { handleInputChange, values, reset } = useForm({
    id,
    title,
    body,
    url,
    createdDate,
  });

  const activeId = useRef(id);

  useMemo(() => {
    if (activeId.current !== id) {
      reset({ title, body, url });
      activeId.current = id;
    }
  }, [body, id, reset, title, url]);

  useEffect(() => {
    dispatch(activeNote(activeId.current, { ...values, url }));
  }, [dispatch, url, values]);

  const handleSaveNote = () => {
    dispatch(startSaveNote({ ...values, id: activeId.current, url }));
  };

  const handleDelete = async () => {
    const { isConfirmed } = await messageDelete();
    if (isConfirmed) {
      dispatch(startDeleteNote(activeId.current));
    }
  };

  return (
    <div className="notes__main-content  animate__animated animate__fadeIn animate_faster">
      <NotesAppBar createdDate={createdDate} handleSaveNote={handleSaveNote} />

      <div className="notes__content">
        <input
          type="text"
          name="title"
          value={values.title}
          placeholder={values.title || "Title"}
          className="notes__title-input"
          autoComplete="off"
          onChange={handleInputChange}
        />
        <textarea
          name="body"
          value={values.body}
          placeholder={values.body || "Description"}
          className="notes__textarea"
          onChange={handleInputChange}
        ></textarea>
        {url && (
          <div className="notes__image">
            <img src={url} alt="some img" />
          </div>
        )}
      </div>
      <button className="btn btn-danger" onClick={handleDelete}>
        {DELETE}
      </button>
    </div>
  );
};

NoteScreen.prototype = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  body: PropTypes.string,
  createdDate: PropTypes.string,
  url: PropTypes.string,
};

export default NoteScreen;
