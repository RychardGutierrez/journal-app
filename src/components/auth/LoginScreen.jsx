import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import isEmail from 'validator/lib/isEmail';
import isStrongPassword from 'validator/lib/isStrongPassword';

import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { uiRemoveErrorMessage, uiSetErrorMessage } from '../../actions/ui';
import { messageErrorWrongEmailOrPassword } from '../../common/errorMessage';
import { LOGIN, SIGN_GOOGLE, CREATE_NEW_ACCOUNT } from '../../common/constants';

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { messageError, loading } = useSelector(({ ui }) => ui);

  // TODO remove this init state
  // email: 'lucas@gmail.com',
  // password: '123Lucas@',
  const { values, handleInputChange } = useForm({
    email: '',
    password: '',
  });

  const { email, password } = values;

  const handleLogin = (e) => {
    e.preventDefault();
    isFormValid() && dispatch(startLoginEmailPassword({ email, password }));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  const isFormValid = () => {
    if (!isEmail(email)) {
      dispatch(uiSetErrorMessage(messageErrorWrongEmailOrPassword));
      return false;
    }

    if (password.trim().length === 0 || !isStrongPassword(password)) {
      dispatch(uiSetErrorMessage(messageErrorWrongEmailOrPassword));
      return false;
    }

    messageError && dispatch(uiRemoveErrorMessage());
    return true;
  };

  return (
    <>
      <h3 className="auth__title">{LOGIN}</h3>
      <form
        onSubmit={handleLogin}
        className="animate__animated animate__fadeIn animate_faster"
      >
        {messageError && (
          <div className="auth__alert-error">{messageError}</div>
        )}

        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          {loading ? 'Loading...' : 'Loading'}
        </button>
        <hr />
        <div className="auth__social_networks ">
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>{SIGN_GOOGLE}</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register" className="link">
          {CREATE_NEW_ACCOUNT}
        </Link>
      </form>
    </>
  );
};

export default LoginScreen;
