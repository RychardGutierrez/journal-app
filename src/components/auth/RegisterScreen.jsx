import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

import isEmail from 'validator/lib/isEmail';
import isStrongPassword from 'validator/lib/isStrongPassword';

import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { uiRemoveErrorMessage, uiSetErrorMessage } from '../../actions/ui';
import { ALREADY_REGISTER, REGISTER } from '../../common/constants';
import {
  messageErrorEmail,
  messageErrorName,
  messageErrorPassword,
} from '../../common/errorMessage';
import { useForm } from '../../hooks/useForm';

const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { messageError } = useSelector(({ ui }) => ui);

  // TODO remove this init state
  // name: "Lucas the duck",
  // email: "lucas.duck@gmail.com",
  // password: "123Lucas@",
  // confirmPassword: "123Lucas@",

  const { values, handleInputChange } = useForm({
    name: '',
    email: '',
    password: '',
    confirmPassword: '',
  });

  const { name, email, password, confirmPassword } = values;

  const handleRegister = (e) => {
    e.preventDefault();

    isFormValid() &&
      dispatch(startRegisterWithEmailPasswordName({ email, password, name }));
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(uiSetErrorMessage(messageErrorName));
      return false;
    }

    if (!isEmail(email)) {
      dispatch(uiSetErrorMessage(messageErrorEmail));
      return false;
    }

    if (
      password.trim() !== confirmPassword.trim() ||
      (!isStrongPassword(password) && !isStrongPassword(confirmPassword))
    ) {
      dispatch(uiSetErrorMessage(messageErrorPassword));
      return false;
    }

    messageError && dispatch(uiRemoveErrorMessage());
    return true;
  };

  return (
    <>
      <h3 className="auth__title">{REGISTER}</h3>
      <form
        onSubmit={handleRegister}
        className="animate__animated animate__fadeIn animate_faster"
      >
        {messageError && (
          <div className="auth__alert-error">{messageError}</div>
        )}

        <input
          type="text"
          placeholder="Full Name"
          name="name"
          autoComplete="off"
          className="auth__input"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Confirm password"
          name="confirmPassword"
          className="auth__input"
          value={confirmPassword}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          {REGISTER}
        </button>

        <Link to="/auth/login" className="link mt-5">
          {ALREADY_REGISTER}
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
