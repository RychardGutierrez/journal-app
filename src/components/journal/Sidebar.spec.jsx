import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import Sidebar from './Sidebar';
configure({ adapter: new Adapter() });

jest.mock('../../actions/auth', () => ({
  startLogout: jest.fn(),
}));

jest.mock('../../actions/notes', () => ({
  startNewNote: jest.fn(),
}));

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  auth: {
    uid: '123-test',
    name: 'Lucas',
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    data: {},
    active: null,
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('<Sidebar />', () => {
  const wrapper = mount(
    <Provider store={store}>
      <Sidebar />
    </Provider>
  );

  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  it('should render component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call startLogout', () => {
    wrapper.find('.btn').prop('onClick')({ preventDefault: () => {} });
    expect(startLogout).toHaveBeenCalled();
  });

  it('should call startNewNote', () => {
    wrapper.find('.journal__new-entry').prop('onClick')({
      preventDefault: () => {},
    });
    expect(startNewNote).toHaveBeenCalled();
  });
});
