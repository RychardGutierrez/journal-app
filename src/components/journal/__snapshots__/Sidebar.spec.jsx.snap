// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render component 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-user"
      />
      <span>
         
        Lucas
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
